### ISSUE TITLE
Ubuntu 20.04 Root on ZFS: Pools not imported on the first boot

### ISSUE THREAD

**Issue by lukasz-zaroda** [openzfs/openzfs-docs#137]
[Ubuntu 20.04 Root on ZFS](https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2020.04%20Root%20on%20ZFS.html#qemu-kvm-xen)

Natively encrypted mirror setup.

After rebooting in step 6, initramfs prompt shows up, and pools are not getting imported. It seems that they have to be manually imported there with:

```
zpool import -d /dev/disk/by-id bpool -f
zpool import -d /dev/disk/by-id rpool -f
```
And only then subsequent reboots work fine. I don't know if it has something to do with https://github.com/openzfs/openzfs-docs/issues/136 .

Not the smoothest experience to say the least :/ .

Edit 1: Actually after that it boots into emergency mode, so yeah, the guide just doesn't work for me.
Edit 2: It turns out after second reboot, only bpool was not imported. But after redoing `zpool import -d /dev/disk/by-id bpool -f` and rebooting now it finally booted... Is it some race condition? Do I have a chance of pool net being imported at boot? I'm not sure.

**Comment by pwyde** [comment]
Can confirm this also. I manage to continue boot process by executing the following:

```
zpool import -f rpool
zpool import -f bpool
zfs load-key -a
exit
```

I also have the issue that after performing `apt dist-upgrade --yes` and `apt install --yes ubuntu-server` and a reboot gives me a empty GRUB menu. It is like all the boot entries have disappeared. I have however not investigated this further. Might be easily resolved by executing `update-grub` before performing final reboot (dunno). Although it worries me that the same issue will appear after a normal `apt upgrade`.

**Comment by rlaager** [comment]
@lukasz-zaroda It's hard to say exactly what happened. It's possible that the pool failed to import because it was marked in use by a system with a different hostid, since it didn't export (because from #136 zed was left running). If it's working now, you're probably okay.

For both of you, I'd suggest checking that the /boot filesystem is mounted. Then make sure that /boot/grub has sane contents (i.e. it's not empty or just /boot/grub/grubenv).

**Comment by pwyde** [comment]
Yes, that must be the mistake. I misunderstood the instructions for "Stop zed". I thought that must only be performed if cache does not get updated. Understanding the whole process a bit more now, I realize zed must be stopped regardless at some point.

I will do more testing and investigation regarding my second issue.

Thank you for this documentation by the way. Being an advanced form of installation with ZFS on root it is much appreciated that we as users have a source of instructions that are kept up to date and open sourced. Previously I had to rely one various blog and forum posts that get outdated as OpenZFS gets updated and developed. Thank you!

**Comment by lukasz-zaroda** [comment]
> Yes, that must be the mistake. I misunderstood the instructions for "Stop zed". I thought that must only be performed if cache does not get updated.

Exactly the same here. After exiting chroot you are no longer able to stop zed. Doc should be a bit more clear about that.

**Comment by pwyde** [comment]
Update on my second issue: empty GRUB menu after install and `apt dist-upgrade --yes` and apt install `--yes ubuntu-server`

I booted into Live CD environment and performed steps according to [Rescuing using a Live CD](https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2020.04%20Root%20on%20ZFS.html#rescuing-using-a-live-cd)

I inspected content of `/boot/grub/grub.cfg` and concluded that it was missing all `menuentry` entries. I chrooted into the installation and executed `update-grub` which generated `menuentry` entries in `/boot/grub/grub.cfg`. Performed a reboot which once again dropped me into emergency shell (busybox) with the error message that pools could not be imported. Just as OP mentioned. Once again I had to execute the following:

```
zpool import -f rpool
zpool import -f bpool
zfs load-key -a
exit
```

After this everything works as intended until I reboot the system which once again gives me an empty GRUB menu(!?) Extremely frustrating and confusing. I will nuke and pave the installation once again. If issue persists I think of registering a new issue regarding this behavior.
